manager/logbroker/broker.go

```diff
@@ -0,0 +1,154 @@
+		return grpc.Errorf(codes.InvalidArgument, "log selector must be provided")
+	}
+
+	if len(selector.ServiceIDs) == 0 && len(selector.TaskIDs) == 0 && len(selector.NodeIDs) == 0 {
+		return grpc.Errorf(codes.InvalidArgument, "log selector must not be empty")
+	}
+
+	return nil
+}
+
+// SubscribeLogs creates a log subscription and streams back logs
+func (lb *LogBroker) SubscribeLogs(request *api.SubscribeLogsRequest, stream api.Logs_SubscribeLogsServer) error {
+	ctx := stream.Context()
+
+	if err := validateSelector(request.Selector); err != nil {
+		return err
+	}
+
+	subscription := &api.SubscriptionMessage{
+		ID:       identity.NewID(),
+		Selector: request.Selector,
+		Options:  request.Options,
+	}
+
+	log := log.G(ctx).WithFields(
+		logrus.Fields{
+			"method":          "(*LogBroker).SubscribeLogs",
+			"subscription.id": subscription.ID,
+		},
+	)
+
+	publishCh, publishCancel := lb.broadcaster.CallbackWatch(events.MatcherFunc(func(event events.Event) bool {
+		publish := event.(*api.PublishLogsRequest)
+		return publish.SubscriptionID == subscription.ID
+	}))
+	defer publishCancel()
+
+	lb.subscriptions.Publish(subscription)
+	defer func() {
+		log.Debug("closing subscription")
+		subscription = subscription.Copy()
+		subscription.Close = true
+		lb.subscriptions.Publish(subscription)
+	}()
+
+	for {
+		select {
+		case event := <-publishCh:
+			publish := event.(*api.PublishLogsRequest)
+			// TODO(aluzzardi): This is broken - we shouldn't stop just because
+			// we received one Close - we should wait for all publishers to Close.
+			if request.Options != nil && !request.Options.Follow && publish.Close {
+				return nil
+			}
+			if err := stream.Send(&api.SubscribeLogsMessage{
+				Messages: publish.Messages,
+			}); err != nil {
+				return err
+			}
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-lb.stopped:
+			return nil
+		}
+	}
+}
+
+// ListenSubscriptions returns a stream of matching subscriptions for the current node
+func (lb *LogBroker) ListenSubscriptions(request *api.ListenSubscriptionsRequest, stream api.LogBroker_ListenSubscriptionsServer) error {
+	remote, err := ca.RemoteNode(stream.Context())
+	if err != nil {
+		return err
+	}
+
+	log := log.G(stream.Context()).WithFields(
+		logrus.Fields{
+			"method": "(*LogBroker).ListenSubscriptions",
+			"node":   remote.NodeID,
+		},
+	)
+	subscriptionCh, subscriptionCancel := lb.subscriptions.Watch()
+	defer subscriptionCancel()
+
+	log.Debug("node registered")
+	for {
+		select {
+		case v := <-subscriptionCh:
+			subscription := v.(*api.SubscriptionMessage)
+			if err := stream.Send(subscription); err != nil {
+				log.Error(err)
+				return err
+			}
+		case <-stream.Context().Done():
+			return stream.Context().Err()
+		case <-lb.stopped:
+			return nil
+		}
+	}
+}
+
+// PublishLogs publishes log messages for a given subscription
+func (lb *LogBroker) PublishLogs(ctx context.Context, request *api.PublishLogsRequest) (*api.PublishLogsResponse, error) {
```

Comment:
What's the authorization model for this RPC to prevent malicious workers from injecting logs into another node's stream? Is the idea that `SubscriptionID` is like a secret token?

It probably wouldn't hurt to get the node ID from `ca.RemoteNode(ctx)` and verify that this particular node was either selected directly in the `LogSelector` or has a task or service that was selected. We should also verify that the node ID given in the `LogContext` inside `request` matches the sender's node ID, that the task ID in `LogContext` is assigned to this node, and that the service ID in `LogContext` matches the task's service.

cc @diogomonica